pkg/kubelet/cm/dra/plugin/client.go

```diff
@@ -97,44 +66,28 @@ func (r *draPluginClient) NodePrepareResources(
 	logger.V(4).Info(log("calling NodePrepareResources rpc"), "request", req)
 	defer logger.V(4).Info(log("done calling NodePrepareResources rpc"), "response", resp, "err", err)
 
-	if r.nodeClientCreator == nil {
-		return nil, errors.New("failed to call NodePrepareResources. nodeClientCreator is nil")
-	}
-
-	nodeClient, nodeClientOld, closer, err := r.nodeClientCreator(r.addr)
+	conn, err := r.plugin.getOrCreateGRPCConn()
 	if err != nil {
 		return nil, err
 	}
-	defer closer.Close()
+	nodeClient := drapb.NewNodeClient(conn)
 
 	ctx, cancel := context.WithTimeout(ctx, PluginClientTimeout)
 	defer cancel()
 
+	if r.plugin.getHighestSupportedGRPCApi() != highestAvailableGRPCApi {
+		// we already tried the v1alpha3 API but it isn't available so falling back automatically
+		resp = nodePrepareResourceFallback(ctx, conn, req)
+		return
+	}
+
 	resp, err = nodeClient.NodePrepareResources(ctx, req)
 	if err != nil {
 		status, _ := grpcstatus.FromError(err)
 		if status.Code() == grpccodes.Unimplemented {
 			// Fall back to the older gRPC API.
-			resp = &drapb.NodePrepareResourcesResponse{
-				Claims: make(map[string]*drapb.NodePrepareResourceResponse),
-			}
-			err = nil
-			for _, claim := range req.Claims {
-				respOld, errOld := nodeClientOld.NodePrepareResource(ctx,
-					&drapbv1alpha2.NodePrepareResourceRequest{
-						Namespace:      claim.Namespace,
-						ClaimUid:       claim.Uid,
-						ClaimName:      claim.Name,
-						ResourceHandle: claim.ResourceHandle,
-					})
-				result := &drapb.NodePrepareResourceResponse{}
-				if errOld != nil {
-					result.Error = errOld.Error()
-				} else {
-					result.CDIDevices = respOld.CdiDevices
-				}
-				resp.Claims[claim.Uid] = result
-			}
```

Comment:
Can we move this check out of band from the *actual* `NodePrepareResources()`call? I'd rather have a dedicated `detectVersion()` call (exact name TBD) that can detect the version and set everything else up as part of ` NewDRAPluginClient()`.

For now, this might mean calling NodePrepareResources() with some bogus data that we know will fail just so we can detect the API version, but maybe we figure out a better way to do this in the future and we just reimplement this one function, rather than having to touch a bunch of other parts of the code.

Reply:
ack